libmscore/rendermidi.cpp

```diff
@@ -439,51 +528,441 @@ static void collectMeasureEvents(EventMap* events, Measure* m, Staff* staff, int
                               collectNote(events, channel, note, velocity, tickOffset, staffIdx);
                  }
             }
+      }
 
-      //
-      // collect program changes and controller
-      //
-      for (Segment* s = m->first(SegmentType::ChordRest); s; s = s->next(SegmentType::ChordRest)) {
-            // int tick = s->tick();
-            for (Element* e : s->annotations()) {
-                  if (!e->isStaffTextBase() || e->staffIdx() < firstStaffIdx || e->staffIdx() >= nextStaffIdx)
+//---------------------------------------------------------
+//   changeCCBetween
+//---------------------------------------------------------
+
+static void changeCCBetween(EventMap* events, int stick, int etick, int startExpr, int endExpr, int channel, int controller, VeloChangeMethod changeMethod, int tickOffset)
+      {
+      // Prevent zero-division error, but add single event
+      if (startExpr == endExpr || stick == etick) {
+            NPlayEvent cc11event = NPlayEvent(ME_CONTROLLER, channel, controller, abs(startExpr));
+            events->insert(std::pair<int, NPlayEvent>(stick + tickOffset, cc11event));
+            return;
+            }
+
+      // Ticks to change expression over
+      int exprTicks = etick - stick;
+      int exprDiff = endExpr - startExpr;
+
+      // The lambdas are in the format:
+      //    func(precalculated values, current tick)
+      // The precalculated values are for values that stay constant, involving
```

Comment:
A bit late note but better late than never :)
Instead of creating a vector of precalculated arguments and passing it as an argument to a lambda function you could save the precalculated values in a lambda closure instead. It would be cleaner and probably a bit more efficient.
For example, for the "default" option:
```
preCalculated.push_back(double(exprDiff));
preCalculated.push_back(double(exprTicks));
valueFunction = [](std::vector<double>& pc, int ct) { return int(pc[0] * (double(ct) / pc[1])); };
```
would become
```
const double dExprDiff = double(exprDiff);
const double dExprTicks = double(exprTicks);
valueFunction = [dExprDiff, dExprTicks](int ct) {
      return int(dExprDiff * (double(ct) / dExprTicks));
      };
```

In C++14 this could probably be done with even nicer syntax, something like
```
valueFunction = [dExprDiff = double(exprDiff), dExprTicks = double(exprTicks)](int ct) {
      return int(dExprDiff * (double(ct) / dExprTicks));
      };
```
but we don't currently use it.

Reply:
Feel free to make a PR to implement that if want :)

Is there a performance improvement because of it?

Reply:
I don't think a performance benefit would be significant, unless this code is called really often, it will just make the code a bit cleaner by avoiding an unnecessary vector and using meaningful variables names inside these formulae.

Reply:
the performance benefit is obvious: instead of allocating some arrays on the heap (which will prevent some obvious compiler optimizations, by the way) direct passing of the exprDiff and exprTicks will use stack memory and allows compiler to use some optimizations. Since changeCCBetween is used in a loop, which iterates over a lot of elements, this code will be called a hell lot of times, which means, that on each iteration the will be unnecessary allocations and deallocations of the small arrays.
Also using a reference to the local variable in the lambda closure may lead to some undesired behaviour in some cases (in this particular case it is ok, of course).